databricks_cli/stack/api.py

```diff
@@ -0,0 +1,246 @@
+# Service, or (c) another written agreement between Licensee and Databricks
+# for the use of the APIs.
+#
+# You may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import json
+from datetime import datetime
+import time
+import six
+
+import click
+
+from requests.exceptions import HTTPError
+from databricks_cli.jobs.api import JobsApi
+from databricks_cli.version import version as CLI_VERSION
+from databricks_cli.configure.config import get_profile_from_context, get_config_for_profile
+from databricks_cli.stack.exceptions import ConfigError
+
+DEBUG_MODE = False
+_home = os.path.expanduser('~')
+MS_SEC = 1000
+STACK_STATUS_INSERT = 'deployed'
+
+# Stack Deployment Status Folder- WIP
+# STACK_DIR = os.path.join(_home, 'databricks', 'stacks', 'beta')
+
+# Resource Types
+JOBS_TYPE = 'job'
+
+# Config Outer Fields
+STACK_NAME = 'name'
+STACK_RESOURCES = 'resources'
+STACK_DEPLOYED = 'deployed'
+
+# Resource Fields
+RESOURCE_ID = 'id'
+RESOURCE_TYPE = 'type'
+RESOURCE_PROPERTIES = 'properties'
+
+# Deployed Resource Fields
+RESOURCE_PHYSICAL_ID = 'physical_id'
+RESOURCE_DEPLOY_OUTPUT = 'deploy_output'
+RESOURCE_DEPLOY_TIMESTAMP = 'timestamp'
+
+
+class StackApi(object):
+    def __init__(self, api_client):
+        self.jobs_client = JobsApi(api_client)
+        self.host = "host/"
+        if click.get_current_context(silent=True):
+            profile = get_profile_from_context()
+            config = get_config_for_profile(profile)
+            self.host = config.host
+        self.deployed_resources = {}
+        self.deployed_resource_config = {}
+
+    def _parse_config_file(self, filename):
+        """Parse the json config"""
+        parsed_conf = {}
+        with open(filename, 'r') as f:
+            parsed_conf = json.load(f)
+
+        return parsed_conf
+
+    def _generate_stack_status_path(self, stack_path):
+        stack_path_split = stack_path.split('.')
+        stack_path_split.insert(-1, STACK_STATUS_INSERT)
+        return '.'.join(stack_path_split)
+
+    def _load_deploy_metadata(self, stack_path, save_path=None):
+        parsed_conf = {}
+        default_status_path = self._generate_stack_status_path(stack_path)
+        try:
+            if os.path.exists(default_status_path):
+                with open(default_status_path, 'r') as f:
+                    parsed_conf = json.load(f)
+                click.echo("Using deployment status file at %s" % default_status_path)
+            elif save_path and os.path.exists(save_path):
```

Comment:
It is unclear why people need to supply a status path. Shall we use the default path only so that the API is simpler?

Reply:
yeah maybe for this PR I can remove it. Maybe we can ask for more feedback about it.

Reply:
Once the user provide a customized path, then they need to remember to provide exactly the same path every time they refer to the stack, which seems to be a big overhead which make this feature not so useful.
This use case is not covered in the PRD, and I am still not sure what is the rational behind it.

Reply:
This was added through Xiangrui's feedback in the design review. The default storage option will still be used even when the save_path is given, so the user doesn't need to use the save_path given at the last deploy if they don't delete the default generated status JSON generated in the same directory as the original stack config template JSON.